independent-projects/tools/analytics-common/src/main/java/io/quarkus/analytics/AnalyticsService.java

```diff
@@ -0,0 +1,260 @@
+            sendIdentity(context);
+            Track trackEvent = Track.builder()
+                    .userId(userId.getUuid())
+                    .context(context)
+                    .event(trackEventType)
+                    .properties(TrackProperties.builder()
+                            .appExtensions(createExtensionsPropertyValue(applicationModel))
+                            .build())
+                    .timestamp(Instant.now())
+                    .build();
+            postFutures.add(this.restClient.postTrack(trackEvent));
+            try {
+                FileUtils.overwrite(trackEvent,
+                        Path.of(localBuildDir.getAbsolutePath(), fileLocations.lastTrackFileName()));
+            } catch (IOException e) {
+                log.warn("Failed to write the last analytics file. " + e.getMessage());
+            }
+        }
+    }
+
+    public void cleanup() {
+        try {
+            // complete all. Normally, the queue should have only 1 element.
+            CompletableFuture.allOf(postFutures.toArray(new CompletableFuture[0])).get(TTL, TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
+            log.warn("No analytics sent. " + e.getClass().getName() + ": " +
+                    (e.getMessage() == null ? "(no message)" : e.getMessage()));
+        }
+    }
+
+    List<TrackProperties.AppExtension> createExtensionsPropertyValue(ApplicationModel applicationModel) {
+        return applicationModel.getDependencies().stream()
+                .filter(dep -> dep.isResolved() &&
+                        dep.isFlagSet(TOP_LEVEL_RUNTIME_EXTENSION_ARTIFACT) &&
+                        onlyPublic(dep.getGroupId()))
+                .map(dep -> TrackProperties.AppExtension.builder()
+                        .groupId(dep.getGroupId())
+                        .artifactId(dep.getArtifactId())
+                        .version(dep.getVersion())
+                        .build())
+                .collect(Collectors.toList());
+    }
+
+    void sendIdentity(final Map<String, Object> context) {
+        if (this.userId.isNew()) { // fixme when inactive on 1st call it will not send identity.
+            this.restClient.postIdentity(Identity.builder()
+                    .userId(this.userId.getUuid())
+                    .context(context)
+                    .timestamp(Instant.now())
+                    .build());
+        }
+    }
+
+    Map<String, Object> createContextMap(ApplicationModel applicationModel,
+            Map<String, Object> buildInfo) {
+        WorkspaceModuleId moduleId = applicationModel.getApplicationModule().getId();
+
+        return new ContextBuilder()
+                .mapPair(PROP_APP)
+                .pair(PROP_NAME, hashSHA256(moduleId.getGroupId() + ":" + moduleId.getArtifactId()))
+                .pair(PROP_VERSION, hashSHA256(moduleId.getArtifactId() + ":" + moduleId.getVersion()))
```

Comment:
Wonder whether artifactId is part of the version?

Reply:
To increase the entropy of the hash and make it harder to crack.